Symfony/CS/AbstractPhpdocTypesFixer.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+/*
+ * This file is part of the PHP CS utility.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Symfony\CS;
+
+use Symfony\CS\DocBlock\DocBlock;
+use Symfony\CS\DocBlock\Line;
+use Symfony\CS\Tokenizer\Tokens;
+
+/**
+ * @author Graham Campbell <graham@mineuk.com>
+ */
+abstract class AbstractPhpdocTypesFixer extends AbstractFixer
+{
+    /**
+     * The annotation tags search inside.
+     *
+     * @var string[]
+     */
+    private static $tags = array('param', 'return', 'type', 'var', 'property');
+
+    /**
+     * {@inheritdoc}
+     */
+    public function fix(\SplFileInfo $file, $content)
+    {
+        $tokens = Tokens::fromCode($content);
+
+        foreach ($tokens as $token) {
+            if (!$token->isGivenKind(T_DOC_COMMENT)) {
+                continue;
+            }
+
+            $doc = new DocBlock($token->getContent());
+            $annotations = $doc->getAnnotationsOfType(self::$tags);
+
+            if (empty($annotations)) {
+                continue;
+            }
+
+            foreach ($annotations as $annotation) {
+                $this->fixScalars($doc->getLine($annotation->getStart()), $annotation->getTag()->getName());
+            }
+
+            $token->setContent($doc->getContent());
+        }
+
+        return $tokens->generateCode();
+    }
+
+    /**
+     * Fix scalar types.
+     *
+     * We must be super careful not to modify parts of words.
+     *
+     * @param Line   $line
+     * @param string $tag
+     */
+    private function fixScalars(Line $line, $tag)
+    {
+        $content = $line->getContent();
+        $tagSplit = preg_split('/\s*\@'.$tag.'\s*/', $content);
+        $spaceSplit = preg_split('/\s/', $tagSplit[1]);
+        $usefulContent = $spaceSplit[0];
+
+        if (strpos($usefulContent, '|') !== false) {
+            $newContent = implode('|', self::normalizeTypes(explode('|', $usefulContent)));
+        } else {
+            $newContent = static::normalizeType($usefulContent);
+        }
+
+        if ($newContent !== $usefulContent) {
+            // limiting to 1 replacement to prevent errors like
+            // "integer $integer" being converted to "int $int"
+            // when they should be converted to "int $integer"
+            $line->setContent(preg_replace('/'.preg_quote($usefulContent).'/', $newContent, $content, 1));
+        }
+    }
+
+    /**
+     * Normalize the given types.
+     *
+     * @param string[] $types
+     *
+     * @return string[]
+     */
+    private static function normalizeTypes(array $types)
+    {
+        foreach ($types as $index => $type) {
+            $types[$index] = static::normalizeType($type);
```

Comment:
On first look I was unable to know why the class is abstract, then i found that method which is not defined in base class, but is called there.
So, when I want to override base class I do not know that I must implement that method and what is it interface.

I know, there is a problem to create abstract static method (only reasonable solution is to create it not abstract but throw sth like NotImplementedException.
Also, static methods are big pain when you want to mockup them.

That is why I would made both methods non-static.